File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofStores/AdminModeStore.swift
import Foundation
import Observation
import WagerproofServices
import WagerproofSharedKit

/// `AdminModeStore` mirrors `wagerproof-mobile/contexts/AdminModeContext.tsx`.
///
/// Two pieces of state:
///   1. `isAdmin` — whether the current user has the `admin` role in
///      Supabase (queried via the `has_role(_user_id, _role)` RPC).
///   2. `adminModeEnabled` — whether the admin has toggled the developer
///      tools on in this session. Persisted to App Group user defaults so
///      it survives launch / re-install for the same user.
///
/// Backend contract — byte-identical to RN's `useIsAdmin` hook:
///   `supabase.rpc('has_role', { _user_id: user.id, _role: 'admin' })`
@Observable
@MainActor
public final class AdminModeStore {
    public private(set) var isAdmin: Bool = false
    public private(set) var isCheckingRole: Bool = false
    public private(set) var lastError: String?

    public var adminModeEnabled: Bool {
        didSet {
            AppGroup.defaults.set(adminModeEnabled, forKey: AppGroupKey.adminModeEnabled)
        }
    }

    /// `true` once a role check has completed for the current user. The view
    /// uses this to avoid flashing developer-mode rows for non-admins while
    /// the RPC is in-flight.
    public private(set) var roleResolved: Bool = false

    public init() {
        self.adminModeEnabled = AppGroup.defaults.bool(forKey: AppGroupKey.adminModeEnabled)
    }

    /// Check the admin role for the given user. Called from the auth lifecycle
    /// handler whenever `phase == .authenticated`. Idempotent — re-runs cheaply.
    public func checkRole(for userId: UUID) async {
        isCheckingRole = true
        defer { isCheckingRole = false }
        do {
            let client = await MainSupabase.shared.client
            let response = try await client
                .rpc("has_role", params: HasRoleParams(_user_id: userId, _role: "admin"))
                .execute()
            // Supabase RPC returns the boolean as the body. Decode it directly.
            if let raw = String(data: response.data, encoding: .utf8) {
                let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
                isAdmin = (trimmed == "true")
            } else {
                isAdmin = false
            }
            roleResolved = true
        } catch {
            // Non-fatal: default to non-admin on error. RN does the same.
            isAdmin = false
            roleResolved = true
            lastError = error.localizedDescription
        }
        // Mirror RN: if the user is not (or no longer) an admin, force the
        // local toggle off so the developer drawer doesn't render at all.
        if !isAdmin && adminModeEnabled {
            adminModeEnabled = false
        }
    }

    /// Clear admin state. Called on sign-out.
    public func reset() {
        isAdmin = false
        adminModeEnabled = false
        roleResolved = false
    }

    public func toggleAdminMode() {
        guard isAdmin else { return }
        adminModeEnabled.toggle()
    }

    public var canEnableAdminMode: Bool { isAdmin }

    // MARK: - DEBUG

    #if DEBUG
    public func debugSet(isAdmin: Bool, modeEnabled: Bool = false) {
        self.isAdmin = isAdmin
        self.adminModeEnabled = modeEnabled
        self.roleResolved = true
    }
    #endif

    private struct HasRoleParams: Encodable {
        let _user_id: UUID
        let _role: String
    }
}

File: wagerproof-android-native/core/stores/src/main/kotlin/com/wagerproof/core/stores/AdminModeStore.kt
package com.wagerproof.core.stores

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.wagerproof.core.services.BuildFlags
import com.wagerproof.core.services.SupabaseClients
import com.wagerproof.core.shared.AppGroupKey
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.rpc
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Port of iOS `AdminModeStore.swift` (doc §13.1).
 *
 * Two pieces of state:
 *   1. [isAdmin] — whether the user has the `admin` role in Supabase (queried
 *      via the `has_role(_user_id, _role)` RPC).
 *   2. [adminModeEnabled] — whether the admin toggled developer tools on;
 *      persisted to App Group prefs so it survives launch for the same user.
 *
 * Backend contract (byte-identical to RN): `rpc('has_role', { _user_id, _role: 'admin' })`.
 */
@Stable
class AdminModeStore {
    var isAdmin by mutableStateOf(false); private set
    var isCheckingRole by mutableStateOf(false); private set
    var lastError by mutableStateOf<String?>(null); private set

    /** `true` once a role check has completed — views use it to avoid flashing dev rows mid-RPC. */
    var roleResolved by mutableStateOf(false); private set

    private var _adminModeEnabled by mutableStateOf(
        StorePrefs.appGroup.getBoolean(AppGroupKey.ADMIN_MODE_ENABLED, false),
    )

    /** Developer-tools toggle; setter persists to App Group prefs. */
    var adminModeEnabled: Boolean
        get() = _adminModeEnabled
        set(value) {
            _adminModeEnabled = value
            StorePrefs.appGroup.edit().putBoolean(AppGroupKey.ADMIN_MODE_ENABLED, value).apply()
        }

    /** Check the admin role for the given user. Called when `phase == authenticated`. Idempotent. */
    suspend fun checkRole(userId: String) {
        isCheckingRole = true
        try {
            val result = SupabaseClients.main.postgrest.rpc(
                "has_role",
                buildJsonObject {
                    put("_user_id", userId)
                    put("_role", "admin")
                },
            )
            // Supabase RPC returns the boolean as the raw body. Decode it directly.
            isAdmin = result.data.trim() == "true"
            roleResolved = true
        } catch (e: Throwable) {
            // Non-fatal: default to non-admin on error. RN does the same.
            isAdmin = false
            roleResolved = true
            lastError = e.localizedMessage ?: e.message
        } finally {
            isCheckingRole = false
        }
        // Mirror RN: if not (or no longer) admin, force the local toggle off so the
        // developer drawer doesn't render at all.
        if (!isAdmin && adminModeEnabled) {
            adminModeEnabled = false
        }
    }

    /** Clear admin state. Called on sign-out. */
    fun reset() {
        isAdmin = false
        adminModeEnabled = false
        roleResolved = false
    }

    fun toggleAdminMode() {
        if (!isAdmin) return
        adminModeEnabled = !adminModeEnabled
    }

    val canEnableAdminMode: Boolean get() = isAdmin

    /** NFL/CFB dry-run staging tables — requires both admin role and the Secret Settings toggle. */
    val dryRunPreviewEnabled: Boolean get() = isAdmin && adminModeEnabled

    // MARK: - DEBUG

    fun debugSet(isAdmin: Boolean, modeEnabled: Boolean = false) {
        if (!BuildFlags.isDebugBuild) return
        this.isAdmin = isAdmin
        this.adminModeEnabled = modeEnabled
        this.roleResolved = true
    }
}
